File: src/main/java/aurora/jwt/decoder/internal/JwtPayloadParser.java
package aurora.jwt.decoder.internal;

import static aurora.jwt.common.util.ValidationUtils.checkArgument;

import java.io.IOException;

import com.fasterxml.jackson.databind.ObjectMapper;
import aurora.jwt.common.dto.SecurityContext;
import aurora.jwt.decoder.internal.payload.IdentityDto;
import aurora.jwt.decoder.internal.payload.SecurityContextDto;

public class JwtPayloadParser {
    private final ObjectMapper objectMapper;
    private final aurora.jwt.decoder.internal.SecurityContextConverter securityContextConverter;

    public JwtPayloadParser(ObjectMapper objectMapper, aurora.jwt.decoder.internal.SecurityContextConverter securityContextConverter) {
        this.objectMapper = objectMapper;
        this.securityContextConverter = securityContextConverter;
    }

    public SecurityContext parse(byte[] jwtPayload) throws IOException {
        SecurityContextDto payload =
                objectMapper.readValue(jwtPayload, SecurityContextDto.class);
        validatePayloadJson(payload);
        return securityContextConverter.convert(payload);
    }

    private void validatePayloadJson(SecurityContextDto payload) {
        IdentityDto identity = payload.getIdentity();
        checkArgument(identity != null, "'identity' is not provided");
        checkArgument(identity.getUserId() != null && !identity.getUserId().isEmpty(), "'identity.userId' is not provided");
        checkArgument(identity.getOrganizationId() != null && !identity.getOrganizationId().isEmpty(), "'identity.organizationId' is not provided");
    }


}

File: src/main/kotlin/aurora/jwt/decoder/internal/JwtPayloadParser.kt
package aurora.jwt.decoder.internal

import aurora.jwt.common.dto.SecurityContext
import aurora.jwt.decoder.internal.payload.PreferencesDto
import aurora.jwt.decoder.internal.payload.SecurityContextDto
import com.fasterxml.jackson.databind.ObjectMapper
import java.io.IOException

class JwtPayloadParser(
    private val objectMapper: ObjectMapper,
    private val securityContextConverter: SecurityContextConverter
) {
    @Throws(IOException::class)
    fun parse(jwtPayload: ByteArray): SecurityContext {
        val payload: SecurityContextDto = objectMapper.readValue(jwtPayload, SecurityContextDto::class.java)

        return payload.toSecurityContext()
    }

    private fun SecurityContextDto.toSecurityContext(): SecurityContext {
        val identity = requireNotNull(identity) { "'identity' is not provided" }
        require(!identity.userId.isNullOrEmpty()) {
            "'identity.userId' is not provided"
        }

        require(
            !identity.organizationId.isNullOrEmpty()
        ) {
            "'identity.organizationId' is not provided"
        }

        return securityContextConverter.convert(
            identity,
            preferences ?: PreferencesDto(),
            requireNotNull(authorization) { "'authorization' is not provided" }
        )
    }
}
